### PR TITLE
fix #2991: race condition in OnheapIncrementalIndex#addToFacts

### LINKED ISSUE #2991: Race condition in IncrementalIndex's facts
In OnheapIncrementalIndex::addToFacts(...), when we get a new fact, we will first add the mapping from key to fact's index, then we do aggregate. It will cause query saw some unaggregated metrics with initial value, and sometimes may cause some weird situations. Like if we use LongMax for build but use LongSum for query, then the result may add LongMax's initial value which is LONG.MIN_VALUE.

BTW, do we need to deal with this:
```
      final Integer prev = facts.putIfAbsent(key, rowIndex);
      if (null == prev) {
        numEntries.incrementAndGet();
      } else {
        // We lost a race
        aggs = concurrentGet(prev);
        // Free up the misfire
        concurrentRemove(rowIndex);
        // This is expected to occur ~80% of the time in the worst scenarios
      }
```
I don't find any multi-thread adds for IncremetalIndex.

### PR DESCRIPTION
**Pull request by KurtYoung**

### CODE REVIEW

File: processing/src/test/java/io/druid/segment/incremental/OnheapIncrementalIndexTest.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
+package io.druid.segment.incremental;
```
Comment (fjy) [inline processing/src/test/java/io/druid/segment/incremental/OnheapIncrementalIndexTest.java]: missing header

Comment (KurtYoung) [inline processing/src/test/java/io/druid/segment/incremental/OnheapIncrementalIndexTest.java]: added

File: processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java [hunk 2]
```diff
@@ -179,11 +180,11 @@ protected Integer addToFacts(
         aggs[i] = agg.factorize(
             selectors.get(agg.getName())
         );
+        aggs[i].aggregate();
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java]: doAggregate also has proper exception handling for parseExceptions, looks like we should handle them here also.  
can we call doAggregate(...) with the new aggs array here instead ?

**Comment by nishantmonu51** [comment]
LGTM, 👍 after fixing handling for parse exceptions (https://github.com/druid-io/druid/pull/3002#discussion_r64343356)

**Comment by fjy** [comment]
👍

File: processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 2 above)
Comment (KurtYoung) [inline processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java]: sry for missing that, will fix it.

**Pull request merged**